### PR TITLE
Feature: Dashboard read operations

### PR DESCRIPTION
**Pull request by Fabricevladimir**
## Description
- includes high level methods for dashboard read operations (uses dummy data)

**NOTE:** Merging into `future-proof-code-layout-mock` instead of `develop` because these operations do not block web-app

Task: [APP-589](https://aragonassociation.atlassian.net/browse/APP-589)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder of the package after the [UPCOMING] title and before the latest version.
- [ ] I have tested my code on the test network.

### CODE REVIEW

File: javascript/modules/client/src/client.ts [hunk 1]
```diff
@@ -282,32 +277,93 @@ export class Client extends ClientCore implements IClient {
...
+      .then(gasLimit => {
+        return this.web3.getApproximateGasFee(gasLimit.toBigInt());
+      });
+  }
+
+  //// PRIVATE METHODS METADATA
+
+  private async _getDaoMetadata(daoIdentifier: string): Promise<DaoMetadata> {
+    // TODO: Implement actual fetch logic using subgraph.
+    // Note: it would be nice if the client could be instantiated with dao identifier
```
Comment (githubdoramon) [inline javascript/modules/client/src/client.ts]: Maybe this would be bad if for any reason you need to fetch info from 2 DAOs at the same time

Comment (Fabricevladimir) [reply javascript/modules/client/src/client.ts]: True story, true story

File: javascript/modules/client/README.md [hunk 2]
```diff
@@ -1,5 +1,4 @@
-Aragon JS SDK Client
----
+## Aragon JS SDK Client
```
Comment (brickpop) [inline javascript/modules/client/README.md]: Why this change?

This is the markdown document title, if we start using `#` or `##`, then everything else needs to be `###` or more

File: javascript/modules/client/README.md [hunk 3]
```diff
@@ -180,16 +179,37 @@ for await (const step of client.methods.deposit(depositParams)) {
...
 ### Loading DAO details
 
-- __TODO__:  Name, metadata, installed plugin list
+Handles retrieving DAO metadata using unique identifier.
+
+```ts
+import { Client } from "@aragon/sdk-client";
+
+const client = new Client(context);
+const daoIdentifier = "0x1234..."; // unique identifier; dao name or address
```
Comment (brickpop) [inline javascript/modules/client/README.md]:
```suggestion
const daoIdentifier = "0x1234..."; // unique identifier; dao ENS domain or address
```

File: javascript/modules/client/README.md [hunk 4]
```diff
@@ -180,16 +179,37 @@ for await (const step of client.methods.deposit(depositParams)) {
...
+
+const metadata = await client.methods.getDaoMetadata(daoIdentifier);
+console.log(metadata);
+
+/* 
+{
+   address: "0x1234...",
+   avatar: "http..."
```
Comment (brickpop) [inline javascript/modules/client/README.md]:
```suggestion
   avatar: "http...",
```

File: javascript/modules/client/README.md [hunk 5]
```diff
@@ -180,16 +179,37 @@ for await (const step of client.methods.deposit(depositParams)) {
...
+   createdAt: 1656411653742,
```
Comment (brickpop) [inline javascript/modules/client/README.md]: I would expect this to be a JS `Date()` as opposed to a simple number

File: javascript/modules/client/README.md [hunk 6]
```diff
@@ -180,16 +179,37 @@ for await (const step of client.methods.deposit(depositParams)) {
...
+   description: "This dao...",
+   links: [{label: "Website", url: "http..."}];
+   name: "Abc Dao",
+   packages: ["ERC20VotingPackage"];
```
Comment (brickpop) [inline javascript/modules/client/README.md]: That will probably change in the future, but for now, I would return the registry ID of the plugins installed
```suggestion
   packages: ["0x1234..."],
```

File: javascript/modules/client/README.md [hunk 7]
```diff
@@ -180,16 +179,37 @@ for await (const step of client.methods.deposit(depositParams)) {
...
+   token: {address: '0x123...', decimals: 18, name: "Test Token", symbol: "TKN"};
```
Comment (brickpop) [inline javascript/modules/client/README.md]: The token is no longer something generic to the DAO. A specific plugin like an ERC20 voting plugin may create its own. Probably we'd need feedback from @Rekard0 or @Michael-A-Heuer

File: javascript/modules/client/README.md [hunk 8]
```diff
@@ -249,42 +271,50 @@ for await (
...
 ### Loading the list of members (ERC20)
 
-- __TODO__
+Retrieving all the members of an ERC20 DAO.
+
+```ts
+import { ClientErc20 } from "@aragon/sdk-client";
+
+const client = new ClientERC20(context);
+const members = await client.methods.getDaoMembers();
```
Comment (brickpop) [inline javascript/modules/client/README.md]:
```suggestion
const members = await client.methods.getMembers();
```

File: javascript/modules/client/src/client-erc20.ts [hunk 9]
```diff
@@ -53,6 +53,7 @@ export class ClientErc20 extends ClientCore implements IClientErc20 {
       this._setDaoConfig(address, config),
     setVotingConfig: (address: string, config: VotingConfig) =>
       this._setVotingConfig(address, config),
+    getDaoMembers: (address: string) => this._getDaoMembers(address),
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]:
```suggestion
    getMembers: (address: string) => this._getMembers(address),
```

File: javascript/modules/client/src/client-erc20.ts [hunk 10]
```diff
@@ -163,46 +161,56 @@ export class ClientErc20 extends ClientCore implements IClientErc20 {
...
+    return Promise.resolve({ average: BigInt(0), max: BigInt(0) });
   }
   private _estimateSetVotingConfig(daoAddress: string, config: VotingConfig) {
     // TODO: Unimplemented
-    return Promise.resolve(BigInt(0));
+    return Promise.resolve({ average: BigInt(0), max: BigInt(0) });
+  }
+
+  //// PRIVATE RETRIEVAL HANDLERS
+  private async _getDaoMembers(daoAddress: string) {
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: If we're wrapping the result within a `Promise.resolve()`, there is not need for async yet.

```suggestion
  private _getDaoMembers(daoAddress: string) {
```

File: javascript/modules/client/src/client-erc20.ts [hunk 11]
```diff
@@ -163,46 +161,56 @@ export class ClientErc20 extends ClientCore implements IClientErc20 {
...
+    if (!daoAddress) {
+      throw new Error("A DAO address is needed");
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]:
```suggestion
      throw new Error("Invalid DAO address or ENS");
```

File: javascript/modules/client/src/client-erc20.ts (hunk 10 above)
Comment (brickpop) [reply javascript/modules/client/src/client-erc20.ts]: Also, for consistency, I would expect `daoIdentifier` as stated on the readme (address / ENS)

File: javascript/modules/client/src/client.ts [hunk 12]
```diff
@@ -17,7 +18,7 @@ import { BigNumber } from "@ethersproject/bignumber";
 import { AddressZero } from "@ethersproject/constants";
 import { ClientCore } from "./internal/core";
 import { DaoRole } from "./internal/interfaces/common";
-import { solidityPack } from "ethers/lib/utils";
+import { solidityPack, isAddress } from "ethers/lib/utils";
```
Comment (brickpop) [inline javascript/modules/client/src/client.ts]: I know you didn't introduce this, but could you fetch these functions from `@ethersproject/xxx`  instead?

File: javascript/modules/client/src/client.ts [hunk 13]
```diff
@@ -282,32 +277,93 @@ export class Client extends ClientCore implements IClient {
...
+
+  //// PRIVATE METHODS METADATA
+
+  private async _getDaoMetadata(daoIdentifier: string): Promise<DaoMetadata> {
+    // TODO: Implement actual fetch logic using subgraph.
+    // Note: it would be nice if the client could be instantiated with dao identifier
+    // thereby removing the need to pass in the identifier for all the data retrieval methods
+
+    if (!daoIdentifier) {
+      throw new Error("A DAO identifier is needed");
```
Comment (brickpop) [inline javascript/modules/client/src/client.ts]: Same as above: `Invalid DAO address or ENS` or similar

File: javascript/modules/client/src/client.ts [hunk 14]
```diff
@@ -282,32 +277,93 @@ export class Client extends ClientCore implements IClient {
...
+    return daoInstance.estimateGas
+      .deposit(tokenAddress, amount, reference, override)
+      .then(gasLimit => {
+        return this.web3.getApproximateGasFee(gasLimit.toBigInt());
+      });
+  }
+
+  //// PRIVATE METHODS METADATA
+
+  private async _getDaoMetadata(daoIdentifier: string): Promise<DaoMetadata> {
```
Comment (brickpop) [inline javascript/modules/client/src/client.ts]: Same as above, async not needed yet

File: javascript/modules/client/src/client.ts [hunk 15]
```diff
@@ -282,32 +277,93 @@ export class Client extends ClientCore implements IClient {
...
+        ? {
+            address: daoIdentifier,
+            name:
+              dummyDaoNames[
+                Math.floor(Math.random() * dummyDaoNames.length - 1)
+              ],
+          }
+        : {
+            address: "0x663ac3c648548eb8ccd292b41a8ff829631c846d",
+            name: daoIdentifier,
```
Comment (brickpop) [inline javascript/modules/client/src/client.ts]: We will not be indexing by DAO name. Indexing will be done by address only. If an ENS is provided, the SDK should resolve the ENS name and use that address. I don't understand the point on returning an input parameter itself

File: javascript/modules/client/src/client.ts [hunk 16]
```diff
@@ -282,32 +277,93 @@ export class Client extends ClientCore implements IClient {
...
+          }),
+
+      createdAt: new Date(
+        fromDate + Math.random() * (Date.now() - fromDate)
+      ).getTime(),
```
Comment (brickpop) [inline javascript/modules/client/src/client.ts]: I would return a native `Date`. consumers can always compute the timestamp themselves

File: javascript/modules/client/src/client.ts [hunk 17]
```diff
@@ -282,32 +277,93 @@ export class Client extends ClientCore implements IClient {
...
+        {
+          label: "Website",
+          url: "https://google.com",
+        },
+        {
+          label: "Discord",
+          url: "https://google.com",
+        },
+      ],
+      packages: ["WhitelistPackage", "ERC20VotingPackage"],
```
Comment (brickpop) [inline javascript/modules/client/src/client.ts]: I would return some 32 byte hex string

File: javascript/modules/client/src/client.ts [hunk 18]
```diff
@@ -282,32 +277,93 @@ export class Client extends ClientCore implements IClient {
...
+      token: {
+        address: "0x0000000000000000000000000000000000000000",
+        name: "Ethereum (Canonical)",
+        symbol: "ETH",
+        decimals: 18,
+      },
```
Comment (brickpop) [inline javascript/modules/client/src/client.ts]: Same as above, not sure about an ERC20 token being an opinionated global thing here. A DAO could also have an ERC721 (NFT) governance plugin and/or not have any token at all

Comment (brickpop) [reply javascript/modules/client/src/client.ts]: It should be up to each plugin's client to provide such information

File: javascript/modules/client/src/client.ts [hunk 19]
```diff
@@ -282,32 +277,93 @@ export class Client extends ClientCore implements IClient {
...
+    "0x" +
+    params.plugins
+      .map(entry => {
+        const item = solidityPack(
+          ["uint256", "bytes[]"],
+          [entry.id, entry.data]
+        );
+        return strip0x(item);
+      })
+      .join("");
```
Comment (brickpop) [inline javascript/modules/client/src/client.ts]: Having conflicting code formatters is an issue. I suggest using `deno fmt` (plugin available). Which is yours?

File: javascript/modules/client/src/internal/interfaces/client.ts [hunk 20]
```diff
@@ -8,15 +8,17 @@ export interface IClient extends IClientCore {
...
       where: string,
       who: string,
       role: DaoRole,
-      data: Uint8Array,
+      data: Uint8Array
     ) => Promise<void>;
     /** Deposits ether or an ERC20 token */
     deposit: (params: IDepositParams) => AsyncGenerator<DaoDepositStepValue>;
+    /** Retrieves metadata for DAO with given identifier*/
+    getDaoMetadata: (daoIdentifier: string) => Promise<DaoMetadata>;
```
Comment (brickpop) [inline javascript/modules/client/src/internal/interfaces/client.ts]:
```suggestion
    getMetadata: (daoIdentifier: string) => Promise<DaoMetadata>;
```

File: javascript/modules/client/src/client.ts (hunk 14 above)
Comment (brickpop) [inline javascript/modules/client/src/client.ts]:
```suggestion
  private async _getMetadata(daoIdentifier: string): Promise<DaoMetadata> {
```

File: javascript/modules/client/src/client.ts [hunk 21]
```diff
@@ -35,12 +36,17 @@ export class Client extends ClientCore implements IClient {
     create: (params: ICreateParams) => this._createDao(params),
     /** Deposits ether or an ERC20 token */
     deposit: (params: IDepositParams) => this._deposit(params),
-    /** Checks whether a role is granted by the curren DAO's ACL settings */
+
+    /** Retrieves metadata for DAO with given identifier*/
+    getDaoMetadata: (daoIdentifier: string) =>
+      this._getDaoMetadata(daoIdentifier),
```
Comment (brickpop) [inline javascript/modules/client/src/client.ts]:
```suggestion
    getMetadata: (daoIdentifier: string) =>
      this._getMetadata(daoIdentifier),
```

File: javascript/modules/client/src/internal/interfaces/client.ts [hunk 22]
```diff
@@ -73,3 +75,24 @@ export type DaoDepositStepValue =
...
+
+// DAO DETAILS
+export type DaoResourceLink = { label: string; url: string };
+
+export type DaoToken = {
+  address: string;
+  decimals: number;
+  name: string;
+  symbol: string;
+};
```
Comment (brickpop) [inline javascript/modules/client/src/internal/interfaces/client.ts]: Given the above points, I'm not sure if this should belong to the ERC20 governance plugin

File: javascript/modules/client/README.md (hunk 2 above)
Comment (Fabricevladimir) [reply javascript/modules/client/README.md]: Automatically formatted

File: javascript/modules/client/src/client.ts (hunk 19 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client.ts]: I've prettier and eslint. Those are part of the dev guidelines we set way back when

File: javascript/modules/client/src/client.ts (hunk 18 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client.ts]: Because of the types from subgraph, I was under the impression that every DAO had a "DAO token" which seems to default to ETH if not explicitly defined.

File: javascript/modules/client/src/client.ts (hunk 15 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client.ts]: Jordi, it's _dummy_ data. It's simply a way when the function is called with an address to return the data with the same address. Obviously this will change when implemented, no?

File: javascript/modules/client/src/client.ts (hunk 12 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client.ts]: Heads up. I was unable to find the `solidityPack` method from `@ethersproject` for whatever reason. I've replaced this with the `pack` method from `@ehtersproject/solidity` which I believe is the same method.

File: javascript/modules/client/README.md [hunk 23]
```diff
@@ -180,16 +180,36 @@ for await (const step of client.methods.deposit(depositParams)) {
 
 ### Loading DAO details
 
-- __TODO__:  Name, metadata, installed plugin list
+Handles retrieving DAO metadata using unique identifier.
```
Comment (brickpop) [inline javascript/modules/client/README.md]:
```suggestion
Handles retrieving DAO metadata using its address or ENS domain.
```

File: javascript/modules/client/README.md [hunk 24]
```diff
@@ -180,16 +180,36 @@ for await (const step of client.methods.deposit(depositParams)) {
...
+
+```ts
+import { Client } from "@aragon/sdk-client";
+
+const client = new Client(context);
+const daoIdentifier = "0x1234..."; // unique identifier; dao ENS domain or address
+
+const metadata = await client.methods.getDaoMetadata(daoIdentifier);
```
Comment (brickpop) [inline javascript/modules/client/README.md]:
```suggestion
const daoAddressOrEns = "0x1234..."; // unique identifier; dao ENS domain or address

const metadata = await client.methods.getDaoMetadata(daoAddressOrEns);
```

File: javascript/modules/client/README.md [hunk 25]
```diff
@@ -249,42 +271,50 @@ for await (
...
 ### Loading the list of members (ERC20)
 
-- __TODO__
+Retrieving all the members of an ERC20 DAO.
+
+```ts
+import { ClientErc20 } from "@aragon/sdk-client";
+
+const client = new ClientERC20(context);
+const members = await client.methods.getMembers();
```
Comment (brickpop) [inline javascript/modules/client/README.md]: We are not passing any kind of DAO/plugin address. If the DAO/plugin address is passed via the context, then we should show it. If it is a `getMembers` parameter, we should show it as well

File: javascript/modules/client/src/client-erc20.ts [hunk 26]
```diff
@@ -53,6 +53,7 @@ export class ClientErc20 extends ClientCore implements IClientErc20 {
       this._setDaoConfig(address, config),
     setVotingConfig: (address: string, config: VotingConfig) =>
       this._setVotingConfig(address, config),
+    getMembers: (daoAddressOrEns: string) => this._getMembers(daoAddressOrEns),
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts:55]:
```suggestion
    /** Retrieves the list of Ethereum addresses that can cast weighted votes on the given DAO's proposals */
    getMembers: (daoAddressOrEns: string) => this._getMembers(daoAddressOrEns),
```

File: javascript/modules/client/src/client-erc20.ts [hunk 27]
```diff
@@ -189,20 +187,30 @@ export class ClientErc20 extends ClientCore implements IClientErc20 {
...
+
+    const mockAddresses = [
+      "0x8367dc645e31321CeF3EeD91a10a5b7077e21f70",
+      "0xDA9dfA130Df4dE4673b89022EE50ff26f6EA73Cf",
+      "0xBE0eB53F46cd790Cd13851d5EFf43D12404d33E8",
+      "0x2dB75d8404144CD5918815A44B8ac3f4DB2a7FAf",
+      "0xc1d60f584879f024299DA0F19Cdb47B931E35b53",
+    ];
+
+    return Promise.resolve(mockAddresses.filter(() => Math.random() > 0.4));
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: Not strictly necessary, just FYI adding a small delay can help the UI check that a loading spinner appears while data is not available. Using a simple `setTimeout` could be of use. 

Example:

```ts
return new Promise(resolve => setTimeout(resolve, 1000)).then(() => mockAddresses.filter(() => Math.random() > 0.4));
```

**Review by brickpop — APPROVED** [review]

File: javascript/modules/client/src/client-erc20.ts [hunk 28]
```diff
@@ -195,14 +201,27 @@ export class ClientErc20 extends ClientCore implements IClientErc20 {
     }
 
     const mockAddresses = [
-      "0x8367dc645e31321CeF3EeD91a10a5b7077e21f70",
+      "0x1234...",
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: Why changing to an incomplete address?

File: javascript/modules/client/src/client-erc20.ts [hunk 29]
```diff
@@ -195,14 +201,27 @@ export class ClientErc20 extends ClientCore implements IClientErc20 {
...
+      mockAddresses.filter(() => Math.random() > 0.4)
+    );
+  }
+
+  private _getProposals(daoAddressOrEns: string) {
+    if (!daoAddressOrEns) {
+      throw new Error("Invalid DAO address or ENS");
+    }
+
+    setProposalStatus(MOCK_PROPOSALS);
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: Modifying an object passed by reference is a bit of an anti pattern. Better to have a new variable containing the result, which is a copy of the original. 

At this stage it doesn't matter, but in general, modifying a global object can have unintended side effects elsewhere

File: javascript/modules/client/src/client-erc20.ts [hunk 30]
```diff
@@ -241,3 +260,79 @@ function unwrapWithdrawParams(
     params.reference ?? "",
   ];
 }
+
+/// HELPER FUNCTIONS TODO: move to utils?
+function setProposalStatus(proposals: IErc20VotingProposal[]) {
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: I would:
- call it `getProposalsWithStatus`
- Do a copy of the input data (`const result = []`), then at each iteration `push` an `Object.assign({}, proposal, {status: ...})`  to the results array
- Return the new array

File: javascript/modules/client/src/client-erc20.ts [hunk 31]
```diff
@@ -241,3 +260,79 @@ function unwrapWithdrawParams(
...
+    creator: "0x1234...",
+
+    startDate,
+    endDate: new Date(startDate.getTime() + 7200000),
+    createdAt: new Date(dateWithinThisYear),
+
+    title: "New Founding for Lorex Lab SubDao",
+    summary:
+      "As most community members know, Aragon has strived to deploy its products to more cost-efficient blockchain networks to facilitate interaction.",
+    proposal: "<h1>This is the super important proposal body<h1>",
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: Careful about HTML. We need to decide what we support and what we don't (HTML/markdown/plain). By now, this is an attack vector that allows anyong to inject `<script>` tags

File: javascript/modules/client/src/client-erc20.ts [hunk 32]
```diff
@@ -241,3 +260,79 @@ function unwrapWithdrawParams(
...
+    resources: [{ url: "https://example.com", description: "Example" }],
+
+    executed: false,
+    status: ProposalStatus.PENDING,
+
+    voteId: "0",
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts:338]: Isn't this redundant with the `id` field at the beginning?

File: javascript/modules/client/src/client-erc20.ts [hunk 33]
```diff
@@ -241,3 +260,79 @@ function unwrapWithdrawParams(
...
+    token: {
+      address: "0x1234...",
+      name: "DAO Token",
+      symbol: "DAO",
+      decimals: 18,
+    },
+
+    yea: 3,
+    nay: 1,
+    abstain: 2,
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: Could we have this within a `results: {}` block?

File: javascript/modules/client/src/client-erc20.ts [hunk 34]
```diff
@@ -241,3 +260,79 @@ function unwrapWithdrawParams(
...
+
+    open: true,
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: What does `open` mean? Isn't this redundant with the `status`?

File: javascript/modules/client/src/client-erc20.ts [hunk 35]
```diff
@@ -241,3 +260,79 @@ function unwrapWithdrawParams(
...
+    participationRequiredPct: 30,
+    supportRequiredPct: 52,
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: Shouldn't this be part of a `config` block maybe?

File: javascript/modules/client/src/client-erc20.ts [hunk 36]
```diff
@@ -241,3 +260,79 @@ function unwrapWithdrawParams(
...
+    votingPower: 135,
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts:361]: Voting power of who?
Is that the census size?

File: javascript/modules/client/src/client-erc20.ts [hunk 37]
```diff
@@ -241,3 +260,79 @@ function unwrapWithdrawParams(
...
+
+    voters: [
+      { id: "0x1334...", voterState: VoteOptions.YEA, weight: 1 },
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: - Does `id` refer to the address of the voter?
- Does `voterState` refer to the vote `value`?

File: javascript/modules/client/README.md [hunk 38]
```diff
@@ -249,42 +273,106 @@ for await (
...
+
+    voteId: "0",
+    participationRequiredPct: 30,
+    supportRequiredPct: 52,
+    votingPower: 135,
+
+    yea: 3,
+    nay: 1,
+    abstain: 2,
+
```
Comment (brickpop) [inline javascript/modules/client/README.md]: Does this reflect the Type defined below?

File: javascript/modules/client/src/client-erc20.ts [hunk 39]
```diff
@@ -233,3 +260,79 @@ function unwrapWithdrawParams(
...
+/// MOCK DATA FOR PROPOSALS
+const dateWithinThisYear = new Date(
+  new Date().setFullYear(new Date().getFullYear() - 1)
+).getTime();
+
+const startDate = new Date(
+  dateWithinThisYear + Math.random() * (Date.now() - dateWithinThisYear)
+);
+
+const MOCK_PROPOSALS = [
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: We should be using the interface to type check the mock data

File: javascript/modules/client/src/client-erc20.ts [hunk 40]
```diff
@@ -233,3 +260,79 @@ function unwrapWithdrawParams(
...
+    endDate: new Date(startDate.getTime() + 7200000),
+    createdAt: new Date(dateWithinThisYear),
+
+    title: "New Founding for Lorex Lab SubDao",
+    summary:
+      "As most community members know, Aragon has strived to deploy its products to more cost-efficient blockchain networks to facilitate interaction.",
+    proposal: "<h1>This is the super important proposal body<h1>",
+    resources: [{ url: "https://example.com", description: "Example" }],
+
+    executed: false,
```
Comment (brickpop) [inline javascript/modules/client/src/client-erc20.ts]: Isn't this redundant, given the `status`?

File: javascript/modules/client/src/client-erc20.ts (hunk 28 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client-erc20.ts]: Copy pasta probably

File: javascript/modules/client/src/internal/interfaces/common.ts [hunk 41]
```diff
@@ -41,3 +41,40 @@ export type GasFeeEstimation = {
   average: bigint;
   max: bigint;
 };
+
+/**
+ * Contains the base structure of a proposal
+ */
+export interface IProposal {
```
Comment (brickpop) [inline javascript/modules/client/src/internal/interfaces/common.ts]: - IMO this should be a strict type, not a flexible interface
- We should use it in the code
- Let's make to mirror the fields defined here correspond to the the logic above

File: javascript/modules/client/src/internal/interfaces/plugins.ts [hunk 42]
```diff
@@ -133,3 +159,28 @@ export enum ProposalCreationSteps {
...
+  nay?: number;
+  abstain?: number;
+
+  open: boolean;
+  participationRequiredPct: number;
+  supportRequiredPct: number;
+  votingPower: number;
+
+  voters: { id: string; voterState: VoteOptions; weight: number }[];
+}
```
Comment (brickpop) [inline javascript/modules/client/src/internal/interfaces/plugins.ts]: As stated above, I would recommend using `types` for strict data and do `... = Proposal & { ... }` 

Also, adapt the modifications suggested before in the code

File: javascript/modules/client/src/internal/interfaces/plugins.ts [hunk 43]
```diff
@@ -133,3 +159,28 @@ export enum ProposalCreationSteps {
 export type ProposalCreationStepValue =
   | { key: ProposalCreationSteps.CREATING; txHash: string }
   | { key: ProposalCreationSteps.DONE; proposalId: string };
+
+// PROPOSAL RETRIEVAL
+
+export interface IErc20VotingProposal extends IProposal {
```
Comment (brickpop) [inline javascript/modules/client/src/internal/interfaces/plugins.ts]:
```suggestion
export interface IErc20Proposal extends IProposal {
```

File: javascript/modules/client/src/internal/interfaces/plugins.ts [hunk 44]
```diff
@@ -133,3 +159,28 @@ export enum ProposalCreationSteps {
...
+
+  open: boolean;
+  participationRequiredPct: number;
+  supportRequiredPct: number;
+  votingPower: number;
+
+  voters: { id: string; voterState: VoteOptions; weight: number }[];
+}
+
+export type DaoToken = {
```
Comment (brickpop) [inline javascript/modules/client/src/internal/interfaces/plugins.ts]:
```suggestion
export type Erc20Token = {
```

Plus renaming where applicable

File: javascript/modules/client/src/client-erc20.ts (hunk 32 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client-erc20.ts:338]: No, they're two different fields. This one is from the ERC20Vote entity.

File: javascript/modules/client/src/client-erc20.ts (hunk 34 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client-erc20.ts]: `open` and `executed` both are somewhat redundant. That said I think they should still be included for convenience because they provide a way to check the open and executed status of the proposal much quicker than say something like the following:
```javascript
  if(proposal.status === ProposalStatus.ACTIVE || proposal.status === ProposalStatus.SUCCEEDED) {
     // proposal is open
  }
```

File: javascript/modules/client/src/client-erc20.ts (hunk 37 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client-erc20.ts]: Yes to both. Worth renaming perhaps if that wasn't clear enough

File: javascript/modules/client/src/client-erc20.ts (hunk 40 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client-erc20.ts]: See above comment concerning `open`

File: javascript/modules/client/README.md (hunk 38 above)
Comment (Fabricevladimir) [reply javascript/modules/client/README.md]: Yup

File: javascript/modules/client/src/client-erc20.ts (hunk 31 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client-erc20.ts]: Good point

File: javascript/modules/client/src/client-erc20.ts (hunk 36 above)
Comment (Fabricevladimir) [reply javascript/modules/client/src/client-erc20.ts:361]: ~Apparently the voting power is used to give weight to the votes and is useful especially when more than one token can participate to the vote. So essentially voting power of the token itself.~

This is the total amount of token at the moment of proposal creation.

**Review by brickpop — APPROVED** [review]
